### PR TITLE
chore: event trigger stack and state status change

### PR DESCRIPTION
**Pull request by tyyzqmf**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

(describe what this merge request does)

## Implementation highlights
截屏2023-12-29 19 14 27

1. Control plane create **Rule For SFN**: Listen state machine status change of control plane region
2. Control plane create **SQS Queue**: Listen all region stacks status change by subscription **SNS topic**
3. Create rule to report stack status change to  **SNS topic**. Each pipeline has a corresponding rule and topic, and will delete it when pipeline deleted.
3. Pipeline status get from DDB without call CloudFormation and StepFunction SDK.
4. **Force refresh** pipeline status by call SFN and CFN sdk when user click on the refresh button with the left mouse button with pressing modifier keys (that is, CTRL, ALT, SHIFT, META)

## Test checklist

- [x] add new test cases
- [ ] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [x] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [ ] deploy data modeling
    - [ ] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

**Review by zxkane — CHANGES_REQUESTED** [review]
- Add test cases for new listenStackStatus and listenStateStatus functions
- Could it recover from below use cases? If so, add test cases for them.

1. manually recover from UPDATE_FAILED,
2. manually updating the VERSION tag of stack to make consistent to the one in DDB

File: src/control-plane/backend/event-bus-construct.ts [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    const ruleState = new Rule(this, 'ListenStateStatusChange', {
+      ruleName: `ClickstreamListenStateStatusChange-${getShortIdOfStack(Stack.of(this))}`,
+      description: 'Rule for listen SFN state machine status change',
+      eventPattern: {
+        source: ['aws.states'],
+        detailType: ['Step Functions Execution Status Change'],
+        detail: {
+          stateMachineArn: [props.listenStateMachine.stateMachineArn],
+        },
+        // resources: [`{ "wildcard": "arn:${Aws.PARTITION}:states:${Aws.REGION}:${Aws.ACCOUNT_ID}:execution:${props.listenStateMachine.stateMachineName}:main*" }`],
```
[the remark this answers is not in the document]
Comment (tyyzqmf) [reply src/control-plane/backend/event-bus-construct.ts:106]: Yes, target lambda will filter execution use prefix `main` .

File: src/control-plane/backend/lambda/api/model/pipeline.ts [hunk 2]
```diff
@@ -288,28 +297,40 @@ export class CPipeline {
     }
   }
 
+  private async _createRules() {
+    if (this.pipeline.region === awsRegion) {
+      return;
+    }
+    const cfnRulePatternResourceArn = `arn:aws:cloudformation:${this.pipeline.region}:${awsAccountId}:stack/Clickstream*${this.pipeline.pipelineId}/*`;
+    await createRuleAndAddTargets(
```
[the remark this answers is not in the document]
Comment (tyyzqmf) [reply src/control-plane/backend/lambda/api/model/pipeline.ts:305]: We will delete this rule when pipeline status change to DELETED in rule's target lambda.

File: src/control-plane/backend/event-bus-construct.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (zxkane) [reply src/control-plane/backend/event-bus-construct.ts:106]: Why not excluding other events via the rule? Any technical limitation.

[the remark this answers is not in the document]
Comment (tyyzqmf) [reply src/control-plane/backend/event-bus-construct.ts:106]: CDK not support create rule with wildcard filter.

[the remark this answers is not in the document]
Comment (zxkane) [reply src/control-plane/backend/event-bus-construct.ts:106]: If it's supported by L1, how about using [escape hatches](https://docs.aws.amazon.com/cdk/v2/guide/cfn_layer.html)?

File: src/control-plane/backend/lambda/api/common/utils.ts [hunk 3]
```diff
@@ -944,13 +952,107 @@ function getStackVersion(pipeline: IPipeline, stackType: PipelineStackType): str
...
+      pipeline.stackDetails ?? pipeline.status?.stackDetails, pipeline.templateVersion);
+  }
+  const executionDetail = pipeline.executionDetail ?? pipeline.status?.executionDetail;
+  let status: PipelineStatusType | undefined;
+  status = _getPipelineStatusFromStacks(pipeline, lastAction);
+  if (executionDetail?.status === ExecutionStatus.FAILED ||
+    executionDetail?.status === ExecutionStatus.TIMED_OUT ||
+    executionDetail?.status === ExecutionStatus.ABORTED) {
+    status = PipelineStatusType.FAILED;
+  } else if (executionDetail?.status === ExecutionStatus.RUNNING) {
```
[the remark this answers is not in the document]
Comment (tyyzqmf) [reply src/control-plane/backend/lambda/api/common/utils.ts:1015]: INGs status include: Creating, Updating, Deleting, so it depend on **lastAction** in Line 1005.

File: src/control-plane/backend/event-bus-construct.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (tyyzqmf) [reply src/control-plane/backend/event-bus-construct.ts]: Use CfnRule addOverride to create wildcard filter.

File: src/control-plane/backend/lambda/api/store/aws/events.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  }
+};
+
+export const putRule = async (region: string, name: string, eventPattern: string) => {
+  try {
+    const client = new CloudWatchEventsClient({
+      ...aws_sdk_client_common_config,
+      region,
+    });
+    const command = new PutRuleCommand({
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/store/aws/events.ts:41]: Let's propogate the tags too.

File: src/control-plane/backend/lambda/api/store/aws/sns.ts [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  }
+};
+
+export const createTopic = async (region: string, name: string) => {
+  try {
+    const client = new SNSClient({
+      ...aws_sdk_client_common_config,
+      region,
+    });
+    const command = new CreateTopicCommand({
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/store/aws/sns.ts:39]: Let's propogate the tags too.

File: src/control-plane/backend/lambda/api/store/aws/sns.ts [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+};
+
+
+export const setPermissionForEventRule = async (region: string, topicArn: string) => {
+  try {
+    const eventRulePolicy = {
+      Version: '2008-10-17',
+      Id: '__default_policy_ID',
+      Statement: [
+        {
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/store/aws/sns.ts:78]: Is it required by this policy?

File: src/control-plane/backend/lambda/api/test/api/application.test.ts [hunk 7]
```diff
@@ -129,6 +135,36 @@ describe('Application test', () => {
...
+                        Data: {
+                          Callback: {
+                            BucketName: 'EXAMPLE_BUCKET',
+                            BucketPrefix: '/ingestion',
+                          },
+                          Input: {
+                            Action: 'Create',
+                            Parameters: [],
+                            StackName: 'clickstream-ingestion1',
+                            TemplateURL: 'https://xxx.com',
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/test/api/application.test.ts:155]: Let's use `example.com`

File: src/control-plane/backend/lambda/listen-stack-status/index.ts [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+  const stackId = eventDetail['stack-id'];
+  const stackName = stackId.split('/')[1];
+
+  if (!stackName.startsWith('Clickstream')) {
+    return;
+  }
+  const stackDetail = await describeStack(stackId, region);
+  if (!stackDetail) {
+    logger.error('Failed to describe stack: ', { stackId, region });
+    return;
```
Comment (zxkane) [inline src/control-plane/backend/lambda/listen-stack-status/index.ts:59]: This swallows this event. I prefer to throw the exception to retry it later.

Comment (zxkane) [inline src/control-plane/backend/lambda/listen-stack-status/index.ts:59]: This swallows this event. I prefer to throw the exception to retry it later.

File: src/control-plane/backend/lambda/listen-stack-status/index.ts [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+  if (stackDetail.ParentId) {
+    return;
+  }
+  logger.info('Detail: ', { stackName: stackName, status: stackDetail.StackStatus });
+
+  const pipelineId = getPipelineIdFromStackName(stackName);
+  const pipeline = await getPipeline(pipelineId);
+  if (!pipeline) {
+    logger.error('Failed to get pipeline by pipelineId: ', { pipelineId });
+    return;
```
Comment (zxkane) [inline src/control-plane/backend/lambda/listen-stack-status/index.ts:70]: This swallows this event. I prefer to throw the exception to retry it later.

File: src/control-plane/backend/lambda/listen-stack-status/index.ts [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+      ExpressionAttributeNames: {
+        '#stackDetails': 'stackDetails',
+      },
+      ExpressionAttributeValues: {
+        ':stackDetails': stackDetails,
+      },
+    };
+    await docClient.send(new UpdateCommand(input));
+  } catch (err) {
+    logger.error('Failed to update pipeline stack status: ', { projectId, pipelineId, err });
```
Comment (zxkane) [inline src/control-plane/backend/lambda/listen-stack-status/index.ts:156]: This swallows this event. I prefer to throw the exception to retry it later.

File: src/control-plane/backend/lambda/listen-state-status/index.ts [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+  if (!executionName?.startsWith('main-')) {
+    return;
+  }
+  logger.info('Detail: ', { executionName: eventDetail.name, status: eventDetail.status });
+
+  const pipelineId = executionName.split('-')[1];
+  const pipeline = await getPipeline(pipelineId);
+  if (!pipeline) {
+    logger.error('Failed to get pipeline by pipelineId: ', { pipelineId });
+    return;
```
Comment (zxkane) [inline src/control-plane/backend/lambda/listen-state-status/index.ts:59]: This swallows this event. I prefer to throw the exception to retry it later.

File: src/control-plane/backend/lambda/listen-state-status/index.ts [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+  }
+};
+
+const getTopicArn = (region: string, pipelineId: string) => {
+  const partition = region.startsWith('cn') ? 'aws-cn' : 'aws';
+  const topicName = `${CFN_TOPIC_PREFIX}-${pipelineId}`;
+  return `arn:${partition}:sns:${region}:${process.env.AWS_ACCOUNT_ID}:${topicName}`;
+};
+
+const getPipeline = async (pipelineId: string) => {
```
Comment (zxkane) [inline src/control-plane/backend/lambda/listen-state-status/index.ts:80]: try to share those duplicated method in those Lambda functions

File: src/control-plane/backend/lambda/listen-state-status/index.ts [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+      return records[0];
+    }
+    return;
+  } catch (err) {
+    logger.error('Failed to query pipeline: ', { pipelineId, err });
+    return;
+  }
+};
+
+const updatePipelineStateStatus = async (
```
Comment (zxkane) [inline src/control-plane/backend/lambda/listen-state-status/index.ts:112]: share it

File: src/control-plane/backend/lambda/listen-state-status/index.ts [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+      ExpressionAttributeNames: {
+        '#executionDetail': 'executionDetail',
+      },
+      ExpressionAttributeValues: {
+        ':executionDetail': executionDetail,
+      },
+    };
+    await docClient.send(new UpdateCommand(input));
+  } catch (err) {
+    logger.error('Failed to update pipeline state status: ', { projectId, pipelineId, err });
```
Comment (zxkane) [inline src/control-plane/backend/lambda/listen-state-status/index.ts:137]: This swallows this event. I prefer to throw the exception to retry it later.

File: src/control-plane/backend/lambda/listen-state-status/index.ts [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+            ':d': { BOOL: true },
+            ':s': { S: PipelineStatusType.DELETED },
+          },
+        },
+      });
+    }
+    const params: TransactWriteItemsCommand = new TransactWriteItemsCommand(transactInput);
+    await docClient.send(params);
+  } catch (err) {
+    logger.error('Failed to delete project: ', { projectId, err });
```
Comment (zxkane) [inline src/control-plane/backend/lambda/listen-state-status/index.ts:181]: This swallows this event. I prefer to throw the exception to retry it later.

File: test.template.json [hunk 16]
```diff
@@ -0,0 +1,4339 @@
+{
```
Comment (zxkane) [inline test.template.json:1]: Is it used?

**Review by zxkane — CHANGES_REQUESTED** [review]

File: frontend/src/apis/pipeline.ts [hunk 17]
```diff
@@ -30,17 +30,16 @@ const getPipelineByProject = async (params: {
...
-  if (params.cache) {
-    cacheParam = 'true';
+  let refreshParam = 'false';
+  if (params.refresh) {
+    refreshParam = params.refresh;
   }
   const result: any = await apiRequest(
     'get',
-    `/pipeline/${params.id}?pid=${params.pid}&cache=${cacheParam}`
+    `/pipeline/${params.projectId}?pid=${params.projectId}&refresh=${refreshParam}`
```
[the remark this answers is not in the document]
Comment (tyyzqmf) [reply frontend/src/apis/pipeline.ts:42]: I will fix it in #738

File: test.template.json (hunk 16 above)
Comment (tyyzqmf) [reply test.template.json:1]: deleted.

File: src/control-plane/backend/lambda/api/store/aws/sns.ts (hunk 6 above)
Comment (tyyzqmf) [reply src/control-plane/backend/lambda/api/store/aws/sns.ts:79]: deleted it

File: src/control-plane/backend/lambda/api/store/aws/sns.ts [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+      region,
+    });
+    const command = new SetTopicAttributesCommand({
+      TopicArn: topicArn,
+      AttributeName: 'Policy',
+      AttributeValue: JSON.stringify(eventRulePolicy),
+    });
+    await client.send(command);
+  } catch (error) {
+    console.error('Error in subscribe topic', { error });
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/store/aws/sns.ts:100]: use logger instead of console

**Review by zxkane — CHANGES_REQUESTED** [review]

**Review by zxkane — APPROVED** [review]

**Pull request merged**